test/test_dos.py

```diff
@@ -5437,6 +5437,236 @@ def test_fat_ds3_share_open_twice(self):
+#include <unistd.h>
+
+#define FNAME "FOO.DAT"
+
+unsigned short sharemode(const char *s) {
+  if (strcmp(s, "SH_COMPAT") == 0)
+    return SH_COMPAT;
+
+  if (strcmp(s, "SH_DENYRW") == 0)
+    return SH_DENYRW;
+
+  if (strcmp(s, "SH_DENYWR") == 0)
+    return SH_DENYWR;
+
+  if (strcmp(s, "SH_DENYRD") == 0)
+    return SH_DENYRD;
+
+  if (strcmp(s, "SH_DENYNO") == 0)
+    return SH_DENYNO;
+
+  return 0xff;
+}
+
+unsigned short openmode(const char *s) {
+  if (strcmp(s, "RW") == 0)
+    return O_RDWR;
+  if (s[0] == 'R')
+    return O_RDONLY;
+  if (s[0] == 'W')
+    return O_WRONLY;
+  return 0xff;
+}
+
+int main(int argc, char *argv[]) {
+  int handle;
+  int ret;
+  int primary = -1;
+  unsigned short prismode, priomode;
+
+  if (argc < 5) {
+    printf("FAIL: Missing arguments (primary|secondary) prismode priomode expected\n");
+    return -2;
+  }
+  if (strcmp(argv[1], "primary") == 0)
+    primary = 1;
+  if (strcmp(argv[1], "secondary") == 0)
+    primary = 0;
+  if (primary < 0) {
+    printf("FAIL: Invalid argument (primary|secondary)\n");
+    return -2;
+  }
+
+  prismode = sharemode(argv[2]);
+  if (prismode == 0xff) {
+    printf("FAIL: Invalid argument prismode '%s'\n", argv[2]);
+    return -2;
+  }
+  priomode = openmode(argv[3]);
+  if (priomode == 0xff) {
+    printf("FAIL: Invalid argument priomode '%s'\n", argv[3]);
+    return -2;
+  }
+
+  // expected result is argv[4]
+
+  // Print results in consistent format
+  // FAIL:("SH_DENYNO", "RW", "ALLOW")[secondary denied]
+
+  if (primary) {
+    unsigned short mode = prismode | priomode;
+
+    // create the file
+    ret = _dos_creat(FNAME, _A_NORMAL, &handle);
+    if (ret != 0) {
+      printf("FAIL:('%s', '%s', '%s')[primary create]\n", argv[2], argv[3], argv[4]);
+      return -1;
+    }
+    _dos_close(handle);
+
+    // open with correct flags
+    ret = _dos_open(FNAME, mode, &handle);
+    if (ret != 0) {
+      printf("FAIL:('%s', '%s', '%s')[primary denied]\n", argv[2], argv[3], argv[4]);
+      return -1;
+    }
+//    printf("INFO: primary: File was opened with mode 0x%04x\n", mode);
+
+    // Now start second copy
+    spawnlp(P_WAIT, argv[0], argv[0], "secondary", argv[2], argv[3], argv[4], NULL);
+
+    _dos_close(handle);
+
+  } else { // secondary
+    /* DJPPP says this about unlink()
+      The POSIX specification requires this removal to be delayed until the
+      file is no longer open.
+
+      So I think we need to call DOS function (int21/41) directly.
+
+      DOS 2+ - UNLINK - DELETE FILE
+
+      AH = 41h
+      DS:DX -> ASCIZ filename (no wildcards, but see notes)
+      CL = attribute mask for deletion (server call only, see notes)
+
+      Return:
+        CF clear if successful
+          AX destroyed (DOS 3.3) AL seems to be drive of deleted file
+        CF set on error
+          AX = error code (02h,03h,05h) (see #01680 at AH=59h/BX=0000h)
+    */
+
+    {
+      char *fname = FNAME;
+      uint8_t carry;
+      asm volatile("stc\n"
```

Comment:
You don't need stc; its only when
the func may be unimplemented.

Reply:
Cool I'll remove that then, thanks.